### ISSUE TITLE
Consider `JsonContext.Strings()` (or similar) to get a JSON path

### ISSUE THREAD

**Issue by anthonyrisinger** [xeipuuv/gojsonschema#228]
I can `c := (*ResultErrorFields).Context()` but how to then use it for reliable traversal (keys will contain periods)? I don't see methods to iterate `head` and `tail` and return something like a `[]string`.

Small change I can throw together, but small enough to simply ask too :-D

**Comment by johandorland** [comment]
So the context can't be used directly for the reason you described. There is [JSON pointer](https://tools.ietf.org/html/draft-ietf-appsawg-json-pointer-07) which can be used for this purpose. It has the right escaping mechanics which prevents collisions with key names. Internally Gojsonschema already uses [Gojsonpointer](https://github.com/xeipuuv/gojsonpointer) which provides this functionality as JSON pointer is used within JSON schema. Other libraries like [jsonschema](https://github.com/santhosh-tekuri/jsonschema) already provide JSON pointers to error locations and that's a very sensible thing to do.

In theory it shouldn't be hard to expose the error location in JSON pointer format and then use that JSON pointer to retrieve the actual JSON on which an error occured. This is something that has been on the back of my mind and would be nice to have. There are a couple of places I suspect the reporting of where an error occurred is slightly off, but no one noticed or bothered to report it yet including myself. Ideally I'd like to have this instead of  `Context()`, but that isn't viable due to backwards compatibility reasons

Returning an error location as a `[]string` path is also a possibility, but it puts a burden on the end user to use it to retrieve the JSON that `[]string` points to. With JSON pointer this is much easier. I don't have a particular implementation in mind, but if anyone wants to give it a go, please do!

**Comment by anthonyrisinger** [comment]
Awesome I'll take a look! I use `gojsonpointer` too and have zero problem heading in that very direction :-)

**Comment by johandorland** [comment]
There has been an open issue for a long time #122 to merge `gojsonpointer` and `gojsonreference` in this repository as they're basically only used in this library. If you're going to expose a `gojsonpointer` somewhere it'd be weird if it was a `gojsonpointer.JsonPointer` and it'd be cleaner if it was a `gojsonschema.JsonPointer` instead, so if you're working on this you might as well do that merge also (and you'd be doing my work, so it would be highly appreciated 😄 ).

**Comment by anthonyrisinger** [comment]
I might have to look into this sooner than later, or do a quick fix on my end. I still maintain allowing something like `Strings()` is safe (assumes no special encoding, like pointers, a simple list of literal keys) and delivers immediate value (folks can walk the resource themselves, or easily convert to a jsonpointer) eg. I created a two-pass validator: first pass is inspected for opportunities to auto-correct and may mutate things before second pass.

Semi-related problem from the same overall bucket, I'm having a _suuuper_ hard time producing actionable output for the user without subSchema context, or some mechanism to know what part of the schema failed (maybe set all `x-*` keys from subSchema to error details?). I'm looking at validation errors across 100+ kubernetes manifests that all report `Must not validate the schema (not)` and nothing else :-( (thoughts on this in particular are much appreciated).

I'm attempting to load a custom `errorMessage` stored on the failing subSchema node (similar to https://github.com/epoberezkin/ajv-errors#usage) and finding it quite impossible to make the experience decent for the user. The custom validators and error formatters are useful, but I can't understand enough about the error to customize it.

This bug stems from a pursuit of the above end; I'd love to know if there's a better path!

**Comment by anthonyrisinger** [comment]
To be clear, the system in play is patching upstream k8s schemas with dozens of business rules. Some of them are best (or possibly only, in edge cases) expressed in terms of `not`, `anyOf`, or `oneOf`, and such keywords result in a poor developer experience. Most rules are working fine, or could possibly be re-written, but I think something like "pass through `x-*` keys as error details" is a nice compromise between flexibility and not wanting to leak the subSchema or change the interface. Thoughts @johandorland ?

**Comment by johandorland** [comment]
It's basically impossible to generate good errors for the `not`, `anyOf` and `oneOf` keywords. I explained why for the `not` case in #222, but you can imagine that `anyOf` and `oneOf` are just as bad.

I'm not 100% sure what you mean by passing al `x-*` keys to error details, but if it means putting all fields starting with `x-` of a JSON instance where an error occurred into that error's `err.Details()` I'm not a fan. It clutters up the `err.Details()` immensely and seems like a very weird and specific hack/workaround to me.

I'm more intrigued by AJV's `errorMessage`. I never came across it before (so thanks for putting it to my attention) and it seems like a very decent way for schema authors to inject a useful error message when an error is occurring. Gojsonschema is not AJV however and I'm not sure it would be possible to replace a gojsonschema error by an `errorMessage`, but maybe putting it in `err.Details()` if present seems doable. The only downside to this is that it isn't part of the JSON schema spec and one of the many extras that AJV offers. Maybe this issue could also be tackled by implementing Draft-08's new standardized output, but a Draft-08 implementation is far off and not a near term solution. I'm going to think about this for a bit. Suggestions on this are welcome.

**Comment by anthonyrisinger** [comment]
**tl;dr** I'd love to see either `errorMessage` passed through unchanged (or any other key eg. `x-gojsonschema`, anything) At the end of the day, I only need a way to reliably propagate some opaque blob stored in the failing subSchema to all the errors it produces.

I can accept the limitations on such keywords, I'm only looking for an escape hatch better than what I'm currently doing for "unset" errors (force a detectable error I can parse and mutate later):
```json
{
  "dependencies": {
    "activeDeadlineSeconds": ["$error,unset activeDeadlineSeconds"]
  }
}
```
Or for general purpose messaging:
```json
{
  "if": { },
  "then": {
    "const": "$error Your message goes here!"
  }
}
```
I'm won't detail the (similar) method I use to register mutating callbacks in the schema tree (I create an impossible `const` error with a jsonpointer to the subSchema in play and the function to call) :-)
